src/magellan_vision/scripts/line_detector.py

```diff
@@ -1,211 +1,170 @@
-                    # x_array_left.append([x1,x2])
-                    # y_array_left.append([y1,y2])
-                    cv2.line(crop2, (x1, y1), (x2, y2), (0, 0, 255), 3)
-                    # cv2.circle(crop, (x1, y1),(5), (0,0,255), 3)
-                    # cv2.circle(crop, (x2, y2),(5), (0,0,255), 3)
-                else:  # <-- Otherwise, right group.
-                    # x_array_right.append([x1,x2])
-                    # y_array_right.append([y1,y2])
-                    cv2.line(crop2, (x1, y1), (x2, y2), (0, 255, 0), 3)
-                    # cv2.circle(crop, (x1, y1),(5), (0,255,0), 3)
-                    # cv2.circle(crop, (x2, y2),(5), (0,255,0), 3)
-            circles = crop
-            lines = crop2
-
-    if linesP is None or len(linesP) < 40:
-        skipped_frames += 1
-
-    cv2.imshow('Crop', crop)
-    # linesAndCircles = cv2.addWeighted(lines, 0.5, circles, 0.5,0)
-    # final = cv2.addWeighted(processed, 0.8, linesAndCircles, 0.5,0)
-    # cv2.imshow('Original', frame)
-    # cv2.imshow('Combo', final)
-    # print(skipped_frames)
-
-    # FIXME: what is this crap, please FIX ME
-    # plt.plot(np.unique(x_array_right),
-    #          np.poly1d(np.polyfit(x_array_right, y_array_right, 1))
-    #          (np.unique(x_array_right)))
-
-    # right_line_slope,right_line_intercept = best_fit(np.asarray(x_array_right),np.asarray(y_array_right))
-    # left_line_slope,left_line_intercept = best_fit(np.asarray(x_array_left),np.asarray(y_array_left))
-
-    toc = time.time()
-    delta_t = toc - tic
-    print("FPS: {}".format(1/delta_t))
-
-    key = cv2.waitKey(25)
-    if key == 27:
-        break
-
-video.release()
-cv2.destroyAllWindows
+
+
+# ROS imports
+import rospy
+import std_msgs.msg
+from sensor_msgs.msg import Image  # ROS Image Message
+from geometry_msgs.msg import Point
+from magellan_core.msg import PointArray
+from cv_bridge import CvBridge, CvBridgeError  # Converts b/w OpenCV Image and ROS Image Message
+
+point_arr = PointArray()
+
+
+class PubSubNode(object):
+    def __init__(self):
+        # TODO FIX THIS
+        try:
+            # Global Constant
+            self._VERBOSE = rospy.get_param("verbose")
+        except KeyError as e:
+            rospy.logerr("LineDetector: error in looking up param. {}".format(e))
+            raise
+
+        self._lock = threading.RLock()
+
+        '''Initialize ros publisher, ros subscriber'''
+        self._lines_object = Lines()
+        # topic where we publish
+        self._image_pub = rospy.Publisher("/perception/color/image_processed", Image, queue_size=5)
+        self._point_arr_pub = rospy.Publisher('/perception/detected_points', PointArray, queue_size=5)
+        self._bridge = CvBridge()
+
+        # subscribed Topic
+        self._subscriber = rospy.Subscriber("/camera/color/image_raw", Image, self._callback, queue_size=1)
+        self._image = None
+
+    def run_detects(self):
+        with self._lock:
```

Comment:
Might be worth just grabbing a reference to `self._image` here so we don't keep the lock in `_callback` blocked while we process this image. Right after you unblock `_callback`, you'll get whatever the first image passed to the callback was - probably a stale image when you get to it. 

Probably not a big deal overall, but will get you slightly fresher images.

Reply:
I agree with Lobos, we should probably do that